Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+    Digit::Time startTime = d1.digit.getTime();
+    uint32_t bxStart = startTime.bunchCrossing;
+    Digit::Time stopTime = d2.stopTime;
+    uint32_t bxStop = stopTime.bunchCrossing;
+    // correct for value rollover
+    if (bxStart < bxStop)
+      bxStart += 0x100000;
+
+    uint32_t stopTimeFull = bxStop + (stopTime.sampaTime << 2);
+    uint32_t startTimeFull = bxStart + (startTime.sampaTime << 2);
+    uint32_t timeDiff = startTimeFull - stopTimeFull;
+
+    if (mPrint) {
+      std::cout << "checkDigits: " << d1.digit.getDetID() << "  " << d1.digit.getPadID() << "  " << bxStop << "  " << stopTime.sampaTime
+                << "  " << bxStart << "  " << startTime.sampaTime << "  " << timeDiff << std::endl;
+    }
+
+    // skip if the time difference is not equal to 1 ADC clock cycle
+    if (timeDiff > 8)
+      return false;
+
+    // merge digits
+    d2.digit.setADC(d1.digit.getADC() + d2.digit.getADC());
+    d2.stopTime = d1.stopTime;
+    d1.merged = true;
+    return true;
+  };
+
+  auto& currentBuffer = getCurrentBuffer();
+  auto currentBufId = getCurrentBufId();
+  auto& previousBuffer = getPreviousBuffer();
+  auto previousBufId = getPreviousBufId();
+
+  if (mPrint) {
+    std::cout << "Merging digits in " << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < previousBuffer.digits.size(); i++) {
+    MergerDigit& d1 = previousBuffer.digits[i];
+
+    // skip digits that do not start at the beginning of the time window
+    Digit::Time startTime = d1.digit.getTime();
+    if (startTime.sampaTime != 0) {
+      continue;
+    }
+
+    for (size_t j = 0; j < previousBuffer.digits.size(); j++) {
+      if (i == j) {
+        continue;
+      }
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+
+  // only merge digits from consecutive orbits
+  uint32_t orbit_p = previousBuffer.orbit;
+  uint32_t orbit_c = currentBuffer.orbit;
+  if (mPrint) {
+    std::cout << "orbit_c: " << orbit_c << "  orbit_p: " << orbit_p << std::endl;
+  }
+  if ((orbit_c >= orbit_p) && ((orbit_c - orbit_p) > 1)) {
+    return;
+  }
+
+  if (mPrint) {
+    std::cout << "Merging digits from " << currentBufId << " (orbit=" << currentBuffer.orbit << ") into "
+              << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < currentBuffer.digits.size(); i++) {
+    MergerDigit& d1 = currentBuffer.digits[i];
+
+    // skip digits that do not start at the beginning of the time window
+    Digit::Time startTime = d1.digit.getTime();
+    if (startTime.sampaTime != 0) {
+      continue;
+    }
+
+    for (size_t j = 0; j < previousBuffer.digits.size(); j++) {
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+}
+
+void Merger::setOrbit(int feeId, uint32_t orbit, bool stop)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
```

Comment:
is this check actually needed in all the merger's functions ? shouldn't the feeid be checked upfront somehow ?

Reply:
Not really. I have added it in order to have some extra checks during the debugging. I propose to remove it once we are ready to drop the WIP label.